pkg/ops/ops.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2016 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ops
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/appc/cni/pkg/ns"
+
+	"github.com/vishvananda/netlink"
+)
+
+type NetOps interface {
+	LinkAdd(link netlink.Link) error
+	LinkDel(link netlink.Link) error
+	LinkByIndex(index int) (netlink.Link, error)
+	LinkByName(name string) (netlink.Link, error)
+	LinkSetName(link netlink.Link, name string) error
+	LinkSetUp(link netlink.Link) error
+	LinkSetMaster(link netlink.Link, master *netlink.Bridge) error
+	LinkSetNsFd(link netlink.Link, fd int) error
+	LinkList() ([]netlink.Link, error)
+
+	AddrAdd(link netlink.Link, addr *netlink.Addr) error
+	AddrList(link netlink.Link, family int) ([]netlink.Addr, error)
+
+	GetNS(nspath string) (ns.NetNS, error)
+	NewNS(persist, switchBack bool) (ns.NetNS, error)
+	WithNetNSPath(nspath string, lockThread bool, f func(ns.NetNS) error) error
+}
+
+var instance NetOps
+var mutex sync.Mutex
+
+func newNetOps(testOps bool) NetOps {
```

Comment:
The existing "netlink.<whatever>" pattern is pretty standard and I don't want to change that.  Instead, the transition to mock operations should be as simple and seamless as possible so nobody is confused.  Instead of "netlink." you now type "Net().".  Only in the test cases do you need any special stuff.

The alternative is to create the "NetOps" object in every function that needs it *and* pass that through all CNI method calls, which I think changes too much code everywhere.

Normally the kernel is the NetOps object because it keeps all the netlink state, but now that we need to abstract that we need some other global object in the background to make sure every function call gets the same view of the world.

Hopefully that makes sense?  I just don't think anything other than a singleton pattern works unless you want to radically re-architect the code to pass an object that's effectively a singleton everywhere.